wallet/walletrpc.c

```diff
@@ -1258,3 +1286,158 @@ static const struct json_command unreserveinputs_command = {
 	false
 };
 AUTODATA(json_command, &unreserveinputs_command);
+
+static struct command_result *match_psbt_inputs_to_utxos(struct command *cmd,
+							 struct wally_psbt *psbt,
+							 struct utxo ***utxos)
+{
+	*utxos = tal_arr(cmd, struct utxo *, 0);
+	for (size_t i = 0; i < psbt->tx->num_inputs; i++) {
+		struct utxo *utxo;
+		struct bitcoin_txid txid;
+
+		wally_tx_input_get_txid(&psbt->tx->inputs[i], &txid);
+		utxo = wallet_utxo_get(*utxos, cmd->ld->wallet,
+				       &txid, psbt->tx->inputs[i].index);
+		if (!utxo)
+			continue;
+
+		/* Oops we haven't reserved this utxo yet.
+		 * Let's just go ahead and reserve it now. */
+		if (utxo->status != output_state_reserved) {
+			if (utxo->status != output_state_available) {
+				return command_fail(cmd, LIGHTNINGD,
+						    "Aborting PSBT signing. UTXO %s:%u is already spent",
+						    type_to_string(tmpctx, struct bitcoin_txid,
+								   &utxo->txid),
+						    utxo->outnum);
+			}
+			/* Reserve this utxo before we spend it */
+			if (!wallet_update_output_status(cmd->ld->wallet,
+							 &txid, utxo->outnum,
+							 output_state_available,
+							 output_state_reserved))
+				return command_fail(cmd, LIGHTNINGD,
+						    "Aborting PSBT signing. Unable to reserve UTXO %s:%u",
+						    type_to_string(tmpctx, struct bitcoin_txid,
+								   &utxo->txid),
+						    utxo->outnum);
```

Comment:
If someone has access to the RPC, forcing them to call 'reserveinputs' before 'signpsbt' is very weak security. :)

Let's talk a bit about *why* we do reservations in the first place. The general concept is that it stops the internal wallet from creating other transactions using that UTXO. If you attempt to use a UTXO that hasn't been marked as 'reserved', there's really no problem with this, except that there's the *possibility* another tx might be constructed that includes it.

The way it's currently written, we consider a 'sign attempt' to indicate that a tx would like to use this utxo, so if it's not already marked as reserved, we should reserve it so that no further txs constructed by the wallet will attempt to include it.

Reply:
Good points. My scenario was more that we have some plugin that does some complex collaborative tx construction, and we can trust (or better we need to trust) the plugin to reserve its own UTXOs correctly, but if we just silently sign independent from our own reservation, we may receive a combined PSBT from the other participants, and pass that to `signpsbt` forgetting to check that the PSBT doesn't include our own unreserved UTXOs, thus silently attributing the UTXO to the attacker. It's not much of a safety net to require reservations, but it might just be what prevents this kind of attack.

Reply:
One way to move forward here would be to split this fixup out into a separate PR and merge the rest :_)

Reply:
Oh I see. One way to mitigate this would be, when we receive inputs from peers to add to the tx we're building, to verify that they're not in fact our inputs. Though this is difficult because it might be that that peer is construction a transaction with a peer who's in turn constructing a transaction with us... but in this case the UTXO would have been marked reserved by the other 'us' so it wouldn't be a problem. 

So I guess the check would be 'if this is our utxo and not already marked reserved, cancel the open with the peer', but you'd still do the check upon receiving a utxo from a peer, which, as code, doesn't exist yet.

Reply:
Somehow some of my comments got lost, let me try to reconstruct them here.

My scenario was more like the following: the attacker doesn't have access to the RPC, but is negotiating some a TX that shares inputs from the victim and the attacker. This is mediated through some plugin, that takes care of the coordination, and has access to the RPC. Now the plugin does call `reserveinputs` when it collects the inputs it wants to add to the TX. The attacker pretends to do the same, but claims an input that the victim owns. If we now have the plugin call `signpsbt`, and it papers over the reservation checks, the plugin will have accidentally signed over ownership of the funds to the attacker, and it can now use those funds.

An example would be a dual-funding plugin that doesn't jump through the extra hoop of making sure that the inputs in the funding, that are claimed by the other endpoint, are not in reality its own funds. If we then sign the PSBT we happen to attribute the funds from our UTXOs, but claimed by the counterparty signed over to the initial balance of the counterparty, which is not what we want.

I think we can easily split out this and continue discussing in a separate PR, while keeping the stronger enforcement of reservations in place. That'd allow us to merge this PR, and not keep it pending.

Reply:
Apparently my comments were not lost after all? GH is weird today...

Reply:
> Oh I see. One way to mitigate this would be, when we receive inputs from peers to add to the tx we're building, to verify that they're not in fact our inputs. Though this is difficult because it might be that that peer is construction a transaction with a peer who's in turn constructing a transaction with us... but in this case the UTXO would have been marked reserved by the other 'us' so it wouldn't be a problem.
> 
> So I guess the check would be 'if this is our utxo and not already marked reserved, cancel the open with the peer', but you'd still do the check upon receiving a utxo from a peer, which, as code, doesn't exist yet. (and thus wouldn't impact this code patch)

Right, I'm not saying that we can't work around this issue by pushing some more validation work onto the plugin, but I'm fearful of being to lenient in what we accept, because that's how a plugin author can easily introduce a bug. It's really more of a belts-and-suspenders kind of thing here :-)

Reply:
Ok. I have come around to seeing things your way. This has been removed and the tests have been updated to adjust to this requirement.